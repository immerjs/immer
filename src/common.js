export const NOTHING =
	typeof Symbol !== "undefined"
		? Symbol("immer-nothing")
		: {["immer-nothing"]: true}

export const DRAFTABLE =
	typeof Symbol !== "undefined" && Symbol.for
		? Symbol.for("immer-draftable")
		: "__$immer_draftable"

export const DRAFT_STATE =
	typeof Symbol !== "undefined" && Symbol.for
		? Symbol.for("immer-state")
		: "__$immer_state"

export function isDraft(value) {
	return !!value && !!value[DRAFT_STATE]
}

export function isDraftable(value) {
	if (!value) return false
	return (
		isPlainObject(value) ||
		!!value[DRAFTABLE] ||
		!!value.constructor[DRAFTABLE] ||
		isMap(value) ||
		isSet(value)
	)
}

export function isPlainObject(value) {
	if (!value || typeof value !== "object") return false
	if (Array.isArray(value)) return true
	const proto = Object.getPrototypeOf(value)
	return !proto || proto === Object.prototype
}

export function original(value) {
	if (value && value[DRAFT_STATE]) {
		return value[DRAFT_STATE].base
	}
	// otherwise return undefined
}

// We use Maps as `drafts` for Sets, not Objects
// See proxy.js
export function assignSet(target, override) {
	override.forEach(value => {
		// When we add new drafts we have to remove their originals if present
		const prev = original(value)
		if (prev) target.delete(prev)
		target.add(value)
	})
	return target
}

// We use Maps as `drafts` for Maps, not Objects
// See proxy.js
export function assignMap(target, override) {
	override.forEach((value, key) => target.set(key, value))
	return target
}

export const assign =
	Object.assign ||
	((target, ...overrides) => {
		overrides.forEach(override =>
			Object.keys(override).forEach(key => (target[key] = override[key]))
		)
		return target
	})

export const ownKeys =
	typeof Reflect !== "undefined" && Reflect.ownKeys
		? Reflect.ownKeys
		: typeof Object.getOwnPropertySymbols !== "undefined"
		? obj =>
				Object.getOwnPropertyNames(obj).concat(
					Object.getOwnPropertySymbols(obj)
				)
		: Object.getOwnPropertyNames

export function shallowCopy(base, invokeGetters = false) {
	if (Array.isArray(base)) return base.slice()
	if (isMap(base)) return new Map(base)
	if (isSet(base)) return new Set(base)
	const clone = Object.create(Object.getPrototypeOf(base))
	ownKeys(base).forEach(key => {
		if (key === DRAFT_STATE) {
			return // Never copy over draft state.
		}
		const desc = Object.getOwnPropertyDescriptor(base, key)
		let {value} = desc
		if (desc.get) {
			if (!invokeGetters) {
				throw new Error("Immer drafts cannot have computed properties")
			}
			value = desc.get.call(base)
		}
		if (desc.enumerable) {
			clone[key] = value
		} else {
			Object.defineProperty(clone, key, {
				value,
				writable: true,
				configurable: true
			})
		}
	})
	return clone
}

export function each(obj, iter) {
	if (Array.isArray(obj) || isMap(obj) || isSet(obj)) {
		obj.forEach((entry, index) => iter(index, entry, obj))
	} else {
		ownKeys(obj).forEach(key => iter(key, obj[key], obj))
	}
}

export function isEnumerable(base, prop) {
	const desc = Object.getOwnPropertyDescriptor(base, prop)
	return !!desc && desc.enumerable
}

export function has(thing, prop) {
	return isMap(thing)
		? thing.has(prop)
		: Object.prototype.hasOwnProperty.call(thing, prop)
}

export function get(thing, prop) {
	return isMap(thing) ? thing.get(prop) : thing[prop]
}

export function is(x, y) {
	// From: https://github.com/facebook/fbjs/blob/c69904a511b900266935168223063dd8772dfc40/packages/fbjs/src/core/shallowEqual.js
	if (x === y) {
		return x !== 0 || 1 / x === 1 / y
	} else {
		return x !== x && y !== y
	}
}

export const hasSymbol = typeof Symbol !== "undefined"

export const hasMap = typeof Map !== "undefined"

export function isMap(target) {
	return hasMap && target instanceof Map
}

export const hasSet = typeof Set !== "undefined"

export function isSet(target) {
	return hasSet && target instanceof Set
}

export function makeIterable(next) {
	let self
	return (self = {
		[Symbol.iterator]: () => self,
		next
	})
}

/** Map.prototype.values _-or-_ Map.prototype.entries */
export function iterateMapValues(state, prop, receiver) {
	const isEntries = prop !== "values"
	return () => {
		const iterator = latest(state)[Symbol.iterator]()
		return makeIterable(() => {
			const result = iterator.next()
			if (!result.done) {
				const [key] = result.value
				const value = receiver.get(key)
				result.value = isEntries ? [key, value] : value
			}
			return result
		})
	}
}

export function makeIterateSetValues(createProxy) {
	function iterateSetValues(state, prop) {
		const isEntries = prop === "entries"
		return () => {
			const iterator = latest(state)[Symbol.iterator]()
			return makeIterable(() => {
				const result = iterator.next()
				if (!result.done) {
					const value = wrapSetValue(state, result.value)
					result.value = isEntries ? [value, value] : value
				}
				return result
			})
		}
	}

	function wrapSetValue(state, value) {
		const key = original(value) || value
		let draft = state.drafts.get(key)
		if (!draft) {
			if (state.finalized || !isDraftable(value) || state.finalizing) {
				return value
			}
			draft = createProxy(value, state)
			state.drafts.set(key, draft)
			if (state.modified) {
				state.copy.add(draft)
			}
		}
		return draft
	}

	return iterateSetValues
}

function latest(state) {
	return state.copy || state.base
}

export function clone(obj) {
	if (!isDraftable(obj)) return obj
	if (Array.isArray(obj)) return obj.map(clone)
	if (isMap(obj)) return new Map(obj)
	if (isSet(obj)) return new Set(obj)
	const cloned = Object.create(Object.getPrototypeOf(obj))
	for (const key in obj) cloned[key] = clone(obj[key])
	return cloned
}

export function freeze(obj, deep = false) {
	if (!isDraftable(obj) || isDraft(obj) || Object.isFrozen(obj)) return
	if (isSet(obj)) {
		obj.add = obj.clear = obj.delete = dontMutateFrozenCollections
	} else if (isMap(obj)) {
		obj.set = obj.clear = obj.delete = dontMutateFrozenCollections
	}
	Object.freeze(obj)
	if (deep) each(obj, (_, value) => freeze(value, true))
}

function dontMutateFrozenCollections() {
	throw new Error("This object has been frozen and should not be mutated")
}
