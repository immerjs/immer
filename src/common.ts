export const PROXY_STATE =
    typeof Symbol !== "undefined"
        ? Symbol("immer-proxy-state")
        : "__$immer_state"

export const RETURNED_AND_MODIFIED_ERROR =
    "An immer producer returned a new value *and* modified its draft. Either return a new value *or* modify the draft."

function verifyMinified() {}

const inProduction =
    (typeof process !== "undefined" && process.env.NODE_ENV === "production") ||
    verifyMinified.name !== "verifyMinified"

let autoFreeze = !inProduction
let useProxies = typeof Proxy !== "undefined"

/**
 * Automatically freezes any state trees generated by immer.
 * This protects against accidental modifications of the state tree outside of an immer function.
 * This comes with a performance impact, so it is recommended to disable this option in production.
 * It is by default enabled.
 *
 * @returns {void}
 */
export function setAutoFreeze(enableAutoFreeze) {
    autoFreeze = enableAutoFreeze
}

export function setUseProxies(value) {
    useProxies = value
}

export function getUseProxies() {
    return useProxies
}

export function isProxy(value) {
    return !!value && !!value[PROXY_STATE]
}

export function isProxyable(value) {
    if (!value) return false
    if (typeof value !== "object") return false
    if (Array.isArray(value)) return true
    const proto = Object.getPrototypeOf(value)
    return proto === null || proto === Object.prototype
}

export function freeze(value) {
    if (autoFreeze) {
        Object.freeze(value)
    }
    return value
}

export function shallowCopy(value) {
    if (Array.isArray(value)) return value.slice()
    if (value.__proto__ === undefined)
        return Object.assign(Object.create(null), value)
    return Object.assign({}, value)
}

export function each(value, cb) {
    if (Array.isArray(value)) {
        for (let i = 0; i < value.length; i++) cb(i, value[i])
    } else {
        for (let key in value) cb(key, value[key])
    }
}

export function has(thing, prop) {
    return Object.prototype.hasOwnProperty.call(thing, prop)
}

// given a base object, returns it if unmodified, or return the changed cloned if modified
export function finalize(base) {
    if (isProxy(base)) {
        const state = base[PROXY_STATE]
        if (state.modified === true) {
            if (state.finalized === true) return state.copy
            state.finalized = true
            return finalizeObject(
                useProxies ? state.copy : (state.copy = shallowCopy(base)),
                state
            )
        } else {
            return state.base
        }
    }
    finalizeNonProxiedObject(base)
    return base
}

function finalizeObject(copy, state) {
    const base = state.base
    each(copy, (prop, value) => {
        if (value !== base[prop]) copy[prop] = finalize(value)
    })
    return freeze(copy)
}

function finalizeNonProxiedObject(parent) {
    // If finalize is called on an object that was not a proxy, it means that it is an object that was not there in the original
    // tree and it could contain proxies at arbitrarily places. Let's find and finalize them as well
    if (!isProxyable(parent)) return
    if (Object.isFrozen(parent)) return
    each(parent, (i, child) => {
        if (isProxy(child)) {
            parent[i] = finalize(child)
        } else finalizeNonProxiedObject(child)
    })
    // always freeze completely new data
    freeze(parent)
}

export function verifyReturnValue(returnedValue, proxy, isProxyModified) {
    if (returnedValue !== undefined && returnedValue !== proxy) {
        // something was returned, and it wasn't the proxy itself
        if (isProxyModified)
            throw new Error(
                "An immer producer returned a new value *and* modified its draft. Either return a new value *or* modify the draft."
            )
    }
}

export function is(x, y) {
    // From: https://github.com/facebook/fbjs/blob/c69904a511b900266935168223063dd8772dfc40/packages/fbjs/src/core/shallowEqual.js
    if (x === y) {
        return x !== 0 || 1 / x === 1 / y
    } else {
        return x !== x && y !== y
    }
}
