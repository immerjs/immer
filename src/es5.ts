"use strict"
// @ts-check

import {
    is,
    isProxyable,
    isProxy,
    PROXY_STATE,
    shallowCopy,
    RETURNED_AND_MODIFIED_ERROR,
    each,
    finalize
} from "./common"

const descriptors = {}
let states = null

function createState(parent, proxy, base) {
    return {
        modified: false,
        hasCopy: false,
        parent,
        base,
        proxy,
        copy: undefined,
        finished: false,
        finalizing: false,
        finalized: false
    }
}

function source(state) {
    return state.hasCopy ? state.copy : state.base
}

function get(state, prop) {
    assertUnfinished(state)
    const value = source(state)[prop]
    if (!state.finalizing && value === state.base[prop] && isProxyable(value)) {
        // only create a proxy if the value is proxyable, and the value was in the base state
        // if it wasn't in the base state, the object is already modified and we will process it in finalize
        prepareCopy(state)
        return (state.copy[prop] = createProxy(state, value))
    }
    return value
}

function set(state, prop, value) {
    assertUnfinished(state)
    if (!state.modified) {
        if (is(source(state)[prop], value)) return
        markChanged(state)
        prepareCopy(state)
    }
    state.copy[prop] = value
}

function markChanged(state) {
    if (!state.modified) {
        state.modified = true
        if (state.parent) markChanged(state.parent)
    }
}

function prepareCopy(state) {
    if (state.hasCopy) return
    state.hasCopy = true
    state.copy = shallowCopy(state.base)
}

// creates a proxy for plain objects / arrays
function createProxy(parent, base) {
    const proxy = shallowCopy(base)
    each(base, i => {
        Object.defineProperty(proxy, "" + i, createPropertyProxy("" + i))
    })
    const state = createState(parent, proxy, base)
    createHiddenProperty(proxy, PROXY_STATE, state)
    states.push(state)
    return proxy
}

function createPropertyProxy(prop) {
    return (
        descriptors[prop] ||
        (descriptors[prop] = {
            configurable: true,
            enumerable: true,
            get() {
                return get(this[PROXY_STATE], prop)
            },
            set(value) {
                set(this[PROXY_STATE], prop, value)
            }
        })
    )
}

function assertUnfinished(state) {
    if (state.finished === true)
        throw new Error(
            "Cannot use a proxy that has been revoked. Did you pass an object from inside an immer function to an async process?"
        )
}

// this sounds very expensive, but actually it is not that expensive in practice
// as it will only visit proxies, and only do key-based change detection for objects for
// which it is not already know that they are changed (that is, only object for which no known key was changed)
function markChanges() {
    // intentionally we process the proxies in reverse order;
    // ideally we start by processing leafs in the tree, because if a child has changed, we don't have to check the parent anymore
    // reverse order of proxy creation approximates this
    for (let i = states.length - 1; i >= 0; i--) {
        const state = states[i]
        if (state.modified === false) {
            if (Array.isArray(state.base)) {
                if (hasArrayChanges(state)) markChanged(state)
            } else if (hasObjectChanges(state)) markChanged(state)
        }
    }
}

function hasObjectChanges(state) {
    const baseKeys = Object.keys(state.base)
    const keys = Object.keys(state.proxy)
    return !shallowEqual(baseKeys, keys)
}

function hasArrayChanges(state) {
    const {proxy} = state
    if (proxy.length !== state.base.length) return true
    // See #116
    // If we first shorten the length, our array interceptors will be removed.
    // If after that new items are added, result in the same original length,
    // those last items will have no intercepting property.
    // So if there is no own descriptor on the last position, we know that items were removed and added
    // N.B.: splice, unshift, etc only shift values around, but not prop descriptors, so we only have to check
    // the last one
    const descriptor = Object.getOwnPropertyDescriptor(proxy, proxy.length - 1)
    // descriptor can be null, but only for newly created sparse arrays, eg. new Array(10)
    if (descriptor && !descriptor.get) return true
    // For all other cases, we don't have to compare, as they would have been picked up by the index setters
    return false
}

export function produceEs5(baseState, producer) {
    const prevStates = states
    states = []
    try {
        // create proxy for root
        const rootProxy = createProxy(undefined, baseState)
        // execute the thunk
        const returnValue = producer.call(rootProxy, rootProxy)
        // and finalize the modified proxy
        each(states, (_, state) => {
            state.finalizing = true
        })
        // find and mark all changes (for parts not done yet)
        // TODO: store states by depth, to be able guarantee processing leaves first
        markChanges()
        let result
        // check whether the draft was modified and/or a value was returned
        if (returnValue !== undefined && returnValue !== rootProxy) {
            // something was returned, and it wasn't the proxy itself
            if (rootProxy[PROXY_STATE].modified)
                throw new Error(RETURNED_AND_MODIFIED_ERROR)
            result = finalize(returnValue)
        } else result = finalize(rootProxy)
        // make sure all proxies become unusable
        each(states, (_, state) => {
            state.finished = true
        })
        return result
    } finally {
        states = prevStates
    }
}

function shallowEqual(objA, objB) {
    //From: https://github.com/facebook/fbjs/blob/c69904a511b900266935168223063dd8772dfc40/packages/fbjs/src/core/shallowEqual.js
    if (is(objA, objB)) return true
    if (
        typeof objA !== "object" ||
        objA === null ||
        typeof objB !== "object" ||
        objB === null
    ) {
        return false
    }
    const keysA = Object.keys(objA)
    const keysB = Object.keys(objB)
    if (keysA.length !== keysB.length) return false
    for (let i = 0; i < keysA.length; i++) {
        if (
            !hasOwnProperty.call(objB, keysA[i]) ||
            !is(objA[keysA[i]], objB[keysA[i]])
        ) {
            return false
        }
    }
    return true
}

function createHiddenProperty(target, prop, value) {
    Object.defineProperty(target, prop, {
        value: value,
        enumerable: false,
        writable: true
    })
}
