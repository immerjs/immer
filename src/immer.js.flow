// @flow

export interface Patch {
    op: "replace" | "remove" | "add",
    path: (string|number)[],
    value?: any
}

export type PatchListener = (patches: Patch[], inversePatches: Patch[]) => void

interface IProduce {
    /**
     * Immer takes a state, and runs a function against it.
     * That function can freely mutate the state, as it will create copies-on-write.
     * This means that the original state will stay unchanged, and once the function finishes, the modified state is returned.
     *
     * If the first argument is a function, this is interpreted as the recipe, and will create a curried function that will execute the recipe
     * any time it is called with the current state.
     *
     * @param currentState - the state to start with
     * @param recipe - function that receives a proxy of the current state as first argument and which can be freely modified
     * @param initialState - if a curried function is created and this argument was given, it will be used as fallback if the curried function is called with a state of undefined
     * @returns The next state: a new state, or the current state if nothing was modified
     */
    <S>(
        currentState: S,
        recipe: (draftState: S) => S | void,
        patchListener?: PatchListener
    ): S;
    // curried invocations with inital state
    <S, A, B, C>(
        recipe: (draftState: S, a: A, b: B, c: C) => S | void,
        initialState: S
    ): (currentState: S | void, a: A, b: B, c: C) => S;
    <S, A, B>(
        recipe: (draftState: S, a: A, b: B) => S | void,
        initialState: S
    ): (currentState: S | void, a: A, b: B) => S;
    <S, A>(
        recipe: (draftState: S, a: A) => S | void,
        initialState: S
    ): (currentState: S | void, a: A) => S;
    <S>(
        recipe: (draftState: S) => S | void,
        initialState: S
    ): (currentState: S | void) => S;
    <S>(
        recipe: (draftState: S, ...extraArgs: any[]) => S | void,
        initialState: S
    ): (currentState: S | void, ...extraArgs: any[]) => S;
    // curried invocations without inital state
    <S, A, B, C>(
        recipe: (draftState: S, a: A, b: B, c: C) => S | void
    ): (currentState: S, a: A, b: B, c: C) => S;
    <S, A, B>(
        recipe: (draftState: S, a: A, b: B) => S | void
    ): (currentState: S, a: A, b: B) => S;
    <S, A>(
        recipe: (draftState: S, a: A) => S | void
    ): (currentState: S, a: A) => S;
    <S>(
        recipe: (draftState: S) => S | void
    ): (currentState: S) => S;
    <S>(
        recipe: (draftState: S, ...extraArgs: any[]) => S | void
    ): (currentState: S, ...extraArgs: any[]) => S;
}

declare export var produce: IProduce
declare export default IProduce

declare export var nothing: typeof undefined

declare export var immerable: Symbol

/**
 * Automatically freezes any state trees generated by immer.
 * This protects against accidental modifications of the state tree outside of an immer function.
 * This comes with a performance impact, so it is recommended to disable this option in production.
 * By default it is turned on during local development, and turned off in production.
 */
declare export function setAutoFreeze(autoFreeze: boolean): void

/**
 * Manually override whether proxies should be used.
 * By default done by using feature detection
 */
declare export function setUseProxies(useProxies: boolean): void

declare export function applyPatches<S>(state: S, patches: Patch[]): S

declare export function original<S>(value: S): ?S

declare export function isDraft(value: any): boolean

/**
 * Creates a mutable draft from an (immutable) object / array. 
 * The draft can be modified until `finishDraft` is called
 */
declare export function createDraft<T>(base: T): T

/**
 * Given a draft that was created using `createDraft`, 
 * finalizes the draft into a new immutable object.
 * Optionally a patch-listener can be provided to gather the patches that are needed to construct the object.
 */
declare export function finishDraft<T>(base: T, listener?: PatchListener): T
