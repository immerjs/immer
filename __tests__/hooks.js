"use strict"
import {Immer, setUseProxies} from "../src/index"
import matchers from "expect/build/matchers"

describe("hooks (proxy) -", () => createHookTests(true))
describe("hooks (es5) -", () => createHookTests(false))

function createHookTests(useProxies) {
    let produce, onAssign, onDelete, onCopy

    beforeEach(() => {
        ;({produce, onAssign, onDelete, onCopy} = new Immer({
            autoFreeze: true,
            useProxies,
            onAssign: defuseProxies(jest.fn().mockName("onAssign")),
            onDelete: defuseProxies(jest.fn().mockName("onDelete")),
            onCopy: defuseProxies(jest.fn().mockName("onCopy"))
        }))
    })

    describe("onAssign()", () => {
        useSharedTests(() => onAssign)
        describe("when draft is an object", () => {
            test("assign", () => {
                produce({a: 0, b: 0, c: 0}, s => {
                    s.a++
                    s.c++
                })
                expectCalls(onAssign)
            })
            test("assign (no change)", () => {
                produce({a: 0}, s => {
                    s.a = 0
                })
                expect(onAssign).not.toBeCalled()
            })
            test("delete", () => {
                produce({a: 1}, s => {
                    delete s.a
                })
                expect(onAssign).not.toBeCalled()
            })
            test("nested assignments", () => {
                produce({a: {b: {c: 1, d: 1, e: 1}}}, s => {
                    const {b} = s.a
                    b.c = 2
                    delete b.d
                    b.e = 1 // no-op
                })
                expectCalls(onAssign)
            })
        })
        describe("when draft is an array", () => {
            test("assign", () => {
                produce([1], s => {
                    s[0] = 0
                })
                expectCalls(onAssign)
            })
            test("push", () => {
                produce([], s => {
                    s.push(4)
                })
                expectCalls(onAssign)
            })
            test("pop", () => {
                produce([1], s => {
                    s.pop()
                })
                expect(onAssign).not.toBeCalled()
            })
            test("unshift", () => {
                produce([1], s => {
                    s.unshift(0)
                })
                expectCalls(onAssign)
            })
            test("length = 0", () => {
                produce([1], s => {
                    s.length = 0
                })
                expect(onAssign).not.toBeCalled()
            })
            test("splice (length += 1)", () => {
                produce([1, 2, 3], s => {
                    s.splice(1, 1, 0, 0)
                })
                expectCalls(onAssign)
            })
            test("splice (length += 0)", () => {
                produce([1, 2, 3], s => {
                    s.splice(1, 1, 0)
                })
                expectCalls(onAssign)
            })
            test("splice (length -= 1)", () => {
                produce([1, 2, 3], s => {
                    s.splice(0, 2, 6)
                })
                expectCalls(onAssign)
            })
        })
        describe("when a draft is moved into a new object", () => {
            it("is called in the right order", () => {
                const calls = []
                onAssign.mockImplementation((_, prop) => {
                    calls.push(prop)
                })
                produce({a: {b: 1, c: {}}}, s => {
                    s.a.b = 0
                    s.a.c.d = 1
                    s.x = {y: {z: s.a}}
                    delete s.a
                })
                // Sibling properties use enumeration order, which means new
                // properties come last among their siblings. The deepest
                // properties always come first in their ancestor chain.
                expect(calls).toEqual(["b", "d", "c", "x"])
            })
        })

        if (useProxies) {
            describe("when draft is a Map", () => {
                test("assign", () => {
                    const key1 = {prop: "val1"}
                    const key2 = {prop: "val2"}
                    produce(new Map([["a", 0], [key1, 1], [key2, 2]]), s => {
                        s.set("a", 10)
                        s.set(key1, 11)
                    })
                    expectCalls(onAssign)
                })
                test("assign (no change)", () => {
                    produce(new Map([["a", 0]]), s => {
                        s.set("a", 0)
                    })
                    expect(onAssign).not.toBeCalled()
                })
                test("delete", () => {
                    produce(new Map([["a", 0]]), s => {
                        s.delete("a")
                    })
                    expect(onAssign).not.toBeCalled()
                })
                test("nested assignments", () => {
                    const key1 = {prop: "val1"}
                    produce(
                        new Map([
                            [
                                "a",
                                new Map([
                                    [
                                        key1,
                                        new Map([["b", 1], ["c", 1], ["d", 1]])
                                    ]
                                ])
                            ]
                        ]),
                        s => {
                            const nested = s.get("a").get(key1)
                            nested.set("b", 2)
                            nested.delete("c")
                            nested.set("d", 1) // no-op
                        }
                    )
                    expectCalls(onAssign)
                })
            })
        }
    })

    describe("onDelete()", () => {
        useSharedTests(() => onDelete)
        describe("when draft is an object -", () => {
            test("delete", () => {
                produce({a: 1, b: 1, c: 1}, s => {
                    delete s.a
                    delete s.c
                })
                expectCalls(onDelete)
            })
            test("delete (no change)", () => {
                produce({}, s => {
                    delete s.a
                })
                expect(onDelete).not.toBeCalled()
            })
            test("nested deletions", () => {
                produce({a: {b: {c: 1}}}, s => {
                    delete s.a.b.c
                })
                expectCalls(onDelete)
            })
        })
        describe("when draft is an array -", () => {
            test("pop", () => {
                produce([1], s => {
                    s.pop()
                })
                expectCalls(onDelete)
            })
            test("length = 0", () => {
                produce([1], s => {
                    s.length = 0
                })
                expectCalls(onDelete)
            })
            test("splice (length -= 1)", () => {
                produce([1, 2, 3], s => {
                    s.splice(0, 2, 6)
                })
                expectCalls(onDelete)
            })
        })

        if (useProxies) {
            describe("when draft is a Map -", () => {
                test("delete", () => {
                    const key1 = {prop: "val1"}
                    const key2 = {prop: "val2"}
                    produce(new Map([["a", 0], [key1, 1], [key2, 2]]), s => {
                        s.delete("a")
                        s.delete(key1)
                    })
                    expectCalls(onDelete)
                })
                test("delete (no change)", () => {
                    produce(new Map(), s => {
                        s.delete("a")
                    })
                    expect(onDelete).not.toBeCalled()
                })
                test("nested deletions", () => {
                    const key1 = {prop: "val1"}
                    produce(
                        new Map([
                            ["a", new Map([[key1, new Map([["b", 1]])]])]
                        ]),
                        s => {
                            s.get("a")
                                .get(key1)
                                .delete("b")
                        }
                    )
                    expectCalls(onDelete)
                })
            })
        }
    })

    describe("onCopy()", () => {
        let calls
        beforeEach(() => {
            calls = []
            onCopy.mockImplementation(s => {
                calls.push(s.base)
            })
        })

        useSharedTests(() => onCopy)
        it("is called in the right order for objects", () => {
            const base = {a: {b: {c: 1}}}
            produce(base, s => {
                delete s.a.b.c
            })
            expect(calls).toShallowEqual([base.a.b, base.a, base])
        })

        if (useProxies) {
            it("is called in the right order for Maps", () => {
                const base = new Map([["a", new Map([["b", 0]])]])
                produce(base, s => {
                    s.get("a").delete("b")
                })
                expect(calls).toShallowEqual([base.get("a"), base])
            })
        }
    })

    function useSharedTests(getHook) {
        it("is called before the parent is frozen", () => {
            const hook = getHook()
            hook.mockImplementation(s => {
                // Parent object must not be frozen.
                expect(Object.isFrozen(s.base)).toBeFalsy()
            })
            produce({a: {b: {c: 0}}}, s => {
                if (hook == onDelete) delete s.a.b.c
                else s.a.b.c = 1
            })
            expect(hook).toHaveBeenCalledTimes(hook == onDelete ? 1 : 3)
        })
    }
}

// Produce a snapshot of the hook arguments (minus any draft state).
function expectCalls(hook) {
    expect(
        hook.mock.calls.map(call => {
            return call.slice(1)
        })
    ).toMatchSnapshot()
}

// For defusing draft proxies.
function defuseProxies(fn) {
    return Object.assign((...args) => {
        expect(args[0].finalized).toBeTruthy()
        args[0].draft = args[0].drafts = null
        fn(...args)
    }, fn)
}

expect.extend({
    toShallowEqual(received, expected) {
        const match = matchers.toBe(received, expected)
        return match.pass || !received || typeof received !== "object"
            ? match
            : !Array.isArray(expected) ||
              (Array.isArray(received) && received.length === expected.length)
            ? matchers.toEqual(received, expected)
            : match
    }
})
