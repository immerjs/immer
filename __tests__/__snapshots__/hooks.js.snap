// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hooks (es5) - onAssign() when draft is an array assign 1`] = `
Array [
  Array [
    0,
    0,
  ],
]
`;

exports[`hooks (es5) - onAssign() when draft is an array push 1`] = `
Array [
  Array [
    0,
    4,
  ],
]
`;

exports[`hooks (es5) - onAssign() when draft is an array splice (length += 0) 1`] = `
Array [
  Array [
    1,
    0,
  ],
]
`;

exports[`hooks (es5) - onAssign() when draft is an array splice (length += 1) 1`] = `
Array [
  Array [
    1,
    0,
  ],
  Array [
    2,
    0,
  ],
  Array [
    3,
    3,
  ],
]
`;

exports[`hooks (es5) - onAssign() when draft is an array splice (length -= 1) 1`] = `
Array [
  Array [
    0,
    6,
  ],
  Array [
    1,
    3,
  ],
]
`;

exports[`hooks (es5) - onAssign() when draft is an array unshift 1`] = `
Array [
  Array [
    0,
    0,
  ],
  Array [
    1,
    1,
  ],
]
`;

exports[`hooks (es5) - onAssign() when draft is an object assign 1`] = `
Array [
  Array [
    "a",
    1,
  ],
  Array [
    "c",
    1,
  ],
]
`;

exports[`hooks (es5) - onAssign() when draft is an object nested assignments 1`] = `
Array [
  Array [
    "c",
    2,
  ],
  Array [
    "b",
    Object {
      "c": 2,
      "e": 1,
    },
  ],
  Array [
    "a",
    Object {
      "b": Object {
        "c": 2,
        "e": 1,
      },
    },
  ],
]
`;

exports[`hooks (es5) - onDelete() when draft is an array - length = 0 1`] = `
Array [
  Array [
    0,
  ],
]
`;

exports[`hooks (es5) - onDelete() when draft is an array - pop 1`] = `
Array [
  Array [
    0,
  ],
]
`;

exports[`hooks (es5) - onDelete() when draft is an array - splice (length -= 1) 1`] = `
Array [
  Array [
    2,
  ],
]
`;

exports[`hooks (es5) - onDelete() when draft is an object - delete 1`] = `
Array [
  Array [
    "a",
  ],
  Array [
    "c",
  ],
]
`;

exports[`hooks (es5) - onDelete() when draft is an object - nested deletions 1`] = `
Array [
  Array [
    "c",
  ],
]
`;

exports[`hooks (proxy) - onAssign() when draft is a Map assign 1`] = `
Array [
  Array [
    "a",
    10,
  ],
  Array [
    Object {
      "prop": "val1",
    },
    11,
  ],
]
`;

exports[`hooks (proxy) - onAssign() when draft is a Map nested assignments 1`] = `
Array [
  Array [
    "b",
    2,
  ],
  Array [
    Object {
      "prop": "val1",
    },
    Map {
      "b" => 2,
      "d" => 1,
    },
  ],
  Array [
    "a",
    Map {
      Object {
        "prop": "val1",
      } => Map {
        "b" => 2,
        "d" => 1,
      },
    },
  ],
]
`;

exports[`hooks (proxy) - onAssign() when draft is a Set assign 1`] = `
Array [
  Array [
    "a",
    Set {
      1,
      2,
      3,
      4,
    },
  ],
]
`;

exports[`hooks (proxy) - onAssign() when draft is a Set delete 1`] = `
Array [
  Array [
    "a",
    Set {
      2,
      3,
    },
  ],
]
`;

exports[`hooks (proxy) - onAssign() when draft is a Set nested assignments 1`] = `
Array [
  Array [
    "a",
    Set {
      "a",
      Set {
        1,
        2,
      },
    },
  ],
]
`;

exports[`hooks (proxy) - onAssign() when draft is an array assign 1`] = `
Array [
  Array [
    0,
    0,
  ],
]
`;

exports[`hooks (proxy) - onAssign() when draft is an array push 1`] = `
Array [
  Array [
    0,
    4,
  ],
]
`;

exports[`hooks (proxy) - onAssign() when draft is an array splice (length += 0) 1`] = `
Array [
  Array [
    1,
    0,
  ],
]
`;

exports[`hooks (proxy) - onAssign() when draft is an array splice (length += 1) 1`] = `
Array [
  Array [
    1,
    0,
  ],
  Array [
    2,
    0,
  ],
  Array [
    3,
    3,
  ],
]
`;

exports[`hooks (proxy) - onAssign() when draft is an array splice (length -= 1) 1`] = `
Array [
  Array [
    0,
    6,
  ],
  Array [
    1,
    3,
  ],
]
`;

exports[`hooks (proxy) - onAssign() when draft is an array unshift 1`] = `
Array [
  Array [
    0,
    0,
  ],
  Array [
    1,
    1,
  ],
]
`;

exports[`hooks (proxy) - onAssign() when draft is an object assign 1`] = `
Array [
  Array [
    "a",
    1,
  ],
  Array [
    "c",
    1,
  ],
]
`;

exports[`hooks (proxy) - onAssign() when draft is an object nested assignments 1`] = `
Array [
  Array [
    "c",
    2,
  ],
  Array [
    "b",
    Object {
      "c": 2,
      "e": 1,
    },
  ],
  Array [
    "a",
    Object {
      "b": Object {
        "c": 2,
        "e": 1,
      },
    },
  ],
]
`;

exports[`hooks (proxy) - onDelete() when draft is a Map - delete 1`] = `
Array [
  Array [
    "a",
  ],
  Array [
    Object {
      "prop": "val1",
    },
  ],
]
`;

exports[`hooks (proxy) - onDelete() when draft is a Map - nested deletions 1`] = `
Array [
  Array [
    "b",
  ],
]
`;

exports[`hooks (proxy) - onDelete() when draft is an array - length = 0 1`] = `Array []`;

exports[`hooks (proxy) - onDelete() when draft is an array - pop 1`] = `
Array [
  Array [
    "0",
  ],
]
`;

exports[`hooks (proxy) - onDelete() when draft is an array - splice (length -= 1) 1`] = `
Array [
  Array [
    "2",
  ],
]
`;

exports[`hooks (proxy) - onDelete() when draft is an object - delete 1`] = `
Array [
  Array [
    "a",
  ],
  Array [
    "c",
  ],
]
`;

exports[`hooks (proxy) - onDelete() when draft is an object - nested deletions 1`] = `
Array [
  Array [
    "c",
  ],
]
`;
